caas/kubernetes/clientconfig/plugins.go

```diff
@@ -303,14 +309,50 @@ func getServiceAccountSecret(clientset kubernetes.Interface, saName, namespace s
 	return clientset.CoreV1().Secrets(sa.Namespace).Get(context.TODO(), sa.Secrets[0].Name, metav1.GetOptions{})
 }
 
-func replaceAuthProviderWithServiceAccountAuthData(
+func newK8sConfigForServiceAccount(
 	contextName string,
+	serviceAccount string,
 	config *clientcmdapi.Config,
-	secret *core.Secret,
-) {
-	authName := config.Contexts[contextName].AuthInfo
-	config.AuthInfos[authName] = &clientcmdapi.AuthInfo{
-		ClientCertificateData: secret.Data[core.ServiceAccountRootCAKey],
-		Token:                 string(secret.Data[core.ServiceAccountTokenKey]),
+	sai corev1.ServiceAccountInterface,
+	seci corev1.SecretInterface,
+) (*clientcmdapi.Config, error) {
+	newConfig := config.DeepCopy()
+
+	saAuth, err := authInfoForServiceAccount(serviceAccount, sai, seci)
+	if err != nil {
+		return newConfig, err
+	}
+
+	authName := newConfig.Contexts[contextName].AuthInfo
+	newConfig.AuthInfos[authName] = saAuth
+	return newConfig, nil
+}
+
+func authInfoForServiceAccount(
+	serviceAccount string,
+	sai corev1.ServiceAccountInterface,
+	seci corev1.SecretInterface,
+) (*clientcmdapi.AuthInfo, error) {
+	sa, err := sai.Get(context.TODO(), serviceAccount, metav1.GetOptions{})
```

Comment:
Can we pass in the context, and just create a single context.TODO() in the calling function

Reply:
Done & Done